source/controllers/include/controllers/Controller.hpp

```diff
@@ -0,0 +1,72 @@
+/**
+ * @author Baptiste Busch
+ * @date 2021/02
+ *
+ */
+
+#pragma once
+
+#include "controllers/exceptions/NotImplementedException.hpp"
+#include "state_representation/Parameters/ParameterInterface.hpp"
+#include <list>
+#include <memory>
+
+namespace controllers {
+/**
+ * @class Controller
+ * @brief Abstract class to define a controller either in joint or cartesian spaces
+ * @tparam SIn the input state type of the controller
+ * @tparam SOut the output command type of the controller
+ */
+template <class SIn, class SOut>
+class Controller {
+public:
+  /**
+   * @brief Empty constructor
+   */
+  explicit Controller();
+
+  /**
+   * @brief Compute the command based on the input state
+   * To be redefined based on the actual controller implementation
+   * @param state the input state of the system. This function accept any number of extra arguments.
+   * @return the output command at the input state
+   */
+  virtual SOut compute_command(const SIn& state, ...) const;
+
+  /**
+   * @brief Compute the command based on the desired state and a feedback state
+   * To be redefined based on the actual controller implementation.
+   * This function accept any number of extra arguments.
+   * @param desired_state the desired state of the system.
+   * @param feedback_state the real state of the system as read from feedback loop
+   * @return the output command at the input state
+   */
+  virtual SOut compute_command(const SIn& desired_state, const SIn& feedback_state) const;
+
+  /**
+   * @brief Return a list of all the parameters of the controller
+   * @return the list of parameters
+   */
+  virtual std::list<std::shared_ptr<StateRepresentation::ParameterInterface>> get_parameters() const;
+};
+
+template <class SIn, class SOut>
+Controller<SIn, SOut>::Controller() {}
```

Comment:
Is an explicit constructor necessary because of the templating? Otherwise I would remove the definition and declaration of the constructor if it's just the default constructor.

Reply:
Good question I actually don't think so. Except if we want to specify what are the accepted combinations (see comment below) then we can do that in the source file.